SpatialGDK/Source/SpatialGDKEditorToolbar/Private/SpatialGDKEditorToolbar.cpp

```diff
@@ -1223,49 +1222,59 @@ void FSpatialGDKEditorToolbarModule::GenerateSchema(bool bFullScan)
+	const bool bFullScanRequired = SpatialGDKEditorInstance->FullScanRequired();
 
-	if (SpatialGDKEditorInstance->FullScanRequired())
+	FSpatialGDKEditor::ESchemaGenerationMethod GenerationMethod;
+	FString OnTaskStartMessage;
+	FString OnTaskCompleteMessage;
+	FString OnTaskFailMessage;
+	if (bFullScanRequired || bFullScan)
 	{
-		OnShowTaskStartNotification("Initial Schema Generation");
-
-		if (SpatialGDKEditorInstance->GenerateSchema(FSpatialGDKEditor::FullAssetScan))
-		{
-			OnShowSuccessNotification("Initial Schema Generation completed!");
-		}
-		else
-		{
-			OnShowFailedNotification("Initial Schema Generation failed");
-			bSchemaBuildError = true;
-		}
+		GenerationMethod = FSpatialGDKEditor::FullAssetScan;
+		const TCHAR* RequiredStr = bFullScanRequired ? TEXT(" required") : TEXT("");
+		OnTaskStartMessage = FString::Printf(TEXT("Generating schema (full scan%s)"), RequiredStr);
+		OnTaskCompleteMessage = TEXT("Full schema generation complete");
+		OnTaskFailMessage = TEXT("Full schema generation failed");
 	}
-	else if (bFullScan)
+	else
 	{
-		OnShowTaskStartNotification("Generating Schema (Full)");
+		GenerationMethod = FSpatialGDKEditor::InMemoryAsset;
+		OnTaskStartMessage = TEXT("Generating schema (incremental)");
+		OnTaskCompleteMessage = TEXT("Incremental schema generation completed!");
+		OnTaskFailMessage = TEXT("Incremental schema generation failed");
+	}
+
+	OnShowTaskStartNotification(OnTaskStartMessage);
+	TSharedFuture<bool> SchemaResult = SpatialGDKEditorInstance->GenerateSchema(GenerationMethod).Share();
+	GenerateSchemaResult(SchemaResult, MoveTemp(OnTaskCompleteMessage), MoveTemp(OnTaskFailMessage));
+}
 
-		if (SpatialGDKEditorInstance->GenerateSchema(FSpatialGDKEditor::FullAssetScan))
+void FSpatialGDKEditorToolbarModule::GenerateSchemaResult(TSharedFuture<bool> SchemaResult, FString OnTaskCompleteMessage,
+														  FString OnTaskFailMessage)
+{
+	if (SchemaResult.IsReady())
+	{
+		bool bResult = SchemaResult.Get();
+		if (bResult)
 		{
-			OnShowSuccessNotification("Full Schema Generation completed!");
+			OnShowSuccessNotification(OnTaskCompleteMessage);
 		}
 		else
 		{
-			OnShowFailedNotification("Full Schema Generation failed");
-			bSchemaBuildError = true;
+			OnShowFailedNotification(OnTaskFailMessage);
 		}
 	}
 	else
 	{
-		OnShowTaskStartNotification("Generating Schema (Incremental)");
-
-		if (SpatialGDKEditorInstance->GenerateSchema(FSpatialGDKEditor::InMemoryAsset))
-		{
-			OnShowSuccessNotification("Incremental Schema Generation completed!");
-		}
-		else
-		{
-			OnShowFailedNotification("Incremental Schema Generation failed");
-			bSchemaBuildError = true;
-		}
+		/* Wait for the schema result to become available. */
+		Async(EAsyncExecution::Thread, [this, SchemaResult, OnTaskCompleteMessage = MoveTemp(OnTaskCompleteMessage),
+										OnTaskFailMessage = MoveTemp(OnTaskFailMessage)]() mutable {
+			SchemaResult.Wait(); // Block and wait for task to complete.
```

Comment:
Worth validating nothing blows up if exiting the editor whilst this is in progress

Reply:
No explosions! Seems to be good